ipq40xx: add support for MikroTik hAP ac3

robimarko opened the pull request:
This adds support for the MikroTik RouterBOARD RBD53iG-5HacD2HnD
(hAP ac³), a  indoor dual band, dual-radio 802.11ac
wireless AP with external omnidirectional antennae, USB port, five
10/100/1000 Mbps Ethernet ports and PoE passthrough.

See https://mikrotik.com/product/hap_ac3 for more info.

Specifications:
 - SoC: Qualcomm Atheros IPQ4019
 - RAM: 256 MB
 - Storage: 16 MB NOR + 128 MB NAND
 - Wireless:
   · Built-in IPQ4019 (SoC) 802.11b/g/n 2x2:2, 3 dBi antennae
   · Built-in IPQ4019 (SoC) 802.11a/n/ac 2x2:2, 5.5 dBi antennae
 - Ethernet: Built-in IPQ4019 (SoC, QCA8075) , 5x 1000/100/10 port,
             passive PoE in, PoE passtrough on port 5
- 1x USB Type A port

Installation:
1. Boot the initramfs image via TFTP
2. Run "cat /proc/mtd" and look for "ubi" partition mtd device number, ex. "mtd1"
3. Use ubiformat to remove MikroTik specific UBI volumes
* Detach the UBI partition by running: "ubidetach -d 0"
* Format the partition by running: "ubiformat /dev/mtdN -y"
Replace mtdN with the correct mtd index from step 2.
3. Flash the sysupgrade image using "sysupgrade -n"

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
build: image: add UBIFS kernel packer

This allows packing the kernel into UBIFS like newer
MikroTik NAND devices require.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
scripts: ubinize-image: make rootfs optional

Currently ubinize-image script always expects the
rootfs image to be passed and a volume for it created.

So, to allow only ubinizing a kernel for example which
the MikroTik hAP ac3 and other new NAND devices from
MikroTik require make rootfs an optional parameter like
kernel.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
build: image: explicitly pass --rootfs to append-ubi

Rootfs is now optional in ubinize-image.sh and
requires --rootfs flag instead of just passing the
rootfs image as the argument before ubinize opts.

So, simply add --rootfs flag before the $(IMAGE_ROOTFS).

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
build: image: add command to ubinize the kernel image

Newer NAND devices from MikroTik like the hAP ac3
require the kernel to be packed into UBIFS and then
ubinized.

So, since the ubinize-image.sh script can now ubinize
kernel only as well lets add a command for it.

This now allows calling ubinize-kernel in the kernel
packaging at then end.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
base-files: dont always create kernel UBI volume

Currently nand_upgrade_tar() will pass the kernel length
to nand_upgrade_prepare_ubi() in all cases except for when
the kernel is to be installed in a separate partition as a
binary with the MTD tool.

While this is fine for almost all cases newer MikroTik NAND
devices like hAP ac3 require the kernel to be installed as a
UBIFS packed UBI volume in its own partition.

So, since we have a custom recipe to use ubiformat to flash
the kernel in its partition it makes no sense for sysupgrade
to also install the kernel as a UBI volume in the "ubi"
partition as it only wastes space and will never be used.

So, simply check whether CI_KERNPART is set to "none" and
if so unset the "has_kernel" variable which will in turn
prevent the kernel length from being passed on and then
the kernel UBI volume wont be created for no usefull purpose.

The ath79 MikroTik NAND target has been setting CI_KERNPART
to "none" for a while now altough that was not preventing
the kernel to be installed as UBI volume as well.

Signed-off-by: Robert Marko <robimarko@gmail.com>

adschm commented:
"image-commands:" -> "build: image:"

adschm commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -44,6 +44,7 @@ ALLWIFIBOARDS:= \
 	linksys_mr8300-v0 \
 	luma_wrtq-329acn \
 	mikrotik_hap-ac2 \
+    mikrotik_hap-ac3 \
```
Broken indent.

robimarko replied:
Sorted

adschm commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -45,6 +45,15 @@ engenius,ens620ext)
 	;;
+mikrotik,hap-ac3)
+	ucidef_set_led_netdev "wan" "WAN" "green:wan" "eth1"
+	ucidef_set_led_netdev "ethernet" "ETHERNET" "green:ethernet" "eth1"
```
ethernet and wan are linked to the same?

adschm replied:
I prefer sticking to the vendor behavior. If people want additional triggers based on preference, they should add them themselves.

robimarko replied:
I got no idea how MikroTik triggers this LED at all.

adschm replied:
So you just assigned the LEDs based on what you thought would be good?

adschm replied:
I don't get it? You are setting an ethernet LED in your code? And internet LED is WAN LED? Why not just use the proper names so there is less confusion?

adschm replied:
> "ethernet" in ROS, so I kept that name and triggered it of the WAN as well.

And that's totally arbitrary, so this should be removed.

> In ROS they have some kind of internet detection, probably some kind of ping test to their server based on which that LED is triggered of.

So OEM does not signal pure WAN link state, like it does for LAN ports?

robimarko replied:
> > "ethernet" in ROS, so I kept that name and triggered it of the WAN as well.
> 
> And that's totally arbitrary, so this should be removed.

Ok, fine will drop it.
> 
> > In ROS they have some kind of internet detection, probably some kind of ping test to their server based on which that LED is triggered of.
> 
> So OEM does not signal pure WAN link-state, like it does for LAN ports?
They do via the port LED and then they also signal internet reachability via the ethernet LED

adschm commented on target/linux/ipq40xx/base-files/etc/board.d/03_gpio_switches:
```diff
@@ -18,6 +18,9 @@ cilab,meshpoint-one)
 	ucidef_add_gpio_switch "sim_card_select" "SIM card select" "3" "0"
 	;;
+mikrotik,hap-ac3)
+	ucidef_add_gpio_switch "poe_passtrough" "PoE Passthrough" "452" "0"
```
Same GPIO as in 01_leds?

robimarko replied:
No, its just triggered of the PoE GPIO

adschm commented on target/linux/ipq40xx/patches-5.10/901-arm-boot-add-dts-files.patch:
```diff
@@ -52,6 +52,7 @@ Signed-off-by: John Crispin <john@phrozen.org>
 +	qcom-ipq4019-fritzbox-7530.dtb \
 +	qcom-ipq4019-fritzrepeater-1200.dtb \
 +	qcom-ipq4019-fritzrepeater-3000.dtb \
++   qcom-ipq4019-hap-ac3.dtb \
```
Broken indent.

robimarko replied:
Sorted

robimarko commented on scripts/ubinize-image.sh:
```diff
@@ -66,20 +66,22 @@ ubilayout() {
 
-	vol_id=$(( $vol_id + 1 ))
-	[ "$rootfs_type" = "ubifs" ] || ubivol $vol_id rootfs_data "" 1
+		vol_id=$(( $vol_id + 1 ))
```
Sorry, I dont get it.
I just added the check for rootfs to this code, other is copy/paste.

robimarko committed:
ipq40xx: add support for MikroTik hAP ac3

This adds support for the MikroTik RouterBOARD RBD53iG-5HacD2HnD
(hAP ac³), a  indoor dual band, dual-radio 802.11ac
wireless AP with external omnidirectional antennae, USB port, five
10/100/1000 Mbps Ethernet ports and PoE passthrough.

See https://mikrotik.com/product/hap_ac3 for more info.

Specifications:
 - SoC: Qualcomm Atheros IPQ4019
 - RAM: 256 MB
 - Storage: 16 MB NOR + 128 MB NAND
 - Wireless:
   · Built-in IPQ4019 (SoC) 802.11b/g/n 2x2:2, 3 dBi antennae
   · Built-in IPQ4019 (SoC) 802.11a/n/ac 2x2:2, 5.5 dBi antennae
 - Ethernet: Built-in IPQ4019 (SoC, QCA8075) , 5x 1000/100/10 port,
             passive PoE in, PoE passtrough on port 5
- 1x USB Type A port

Installation:
1. Boot the initramfs image via TFTP
2. Run "cat /proc/mtd" and look for "ubi" partition mtd device number, ex. "mtd1"
3. Use ubiformat to remove MikroTik specific UBI volumes
* Detach the UBI partition by running: "ubidetach -d 0"
* Format the partition by running: "ubiformat /dev/mtdN -y"
Replace mtdN with the correct mtd index from step 2.
3. Flash the sysupgrade image using "sysupgrade -n"

Signed-off-by: Robert Marko <robimarko@gmail.com>

voip2014 reviewed (commented):
The changes are working on the latest commit and its functions as expected

robimarko commented:
@adschm Can we get this merged?
It's been tested thoroughly

lukasmrtvy commented:
Do We need to wait for major release, or is possible to cherry pick it to v21.02.2 ?

garis commented:
Tested-by: Riccardo Gasparini garis94@gmail.com

robimarko commented:
@chunkeey Any chance of you merging this as it's been thoroughly tested?
I will just rebase it to get rid of the conflicts

TrashPandaCodingGarbage commented:
@chunkeey Please merge. Much love <3

tomkins commented:
`Tested-by: Alex Tomkins <tomkins@darkzone.net>`

hauke commented:
Thank you for the patches, I applied them to master in da3261e57cf33d6804222c614c77e13fdb32cf95, 5a305e429facce7fae71f5589b4adbb55b918c9c, 6db4a0372a726ebc1720adf308d698a5df9f1af1, 1fbc9c5e4dea7311236c8f74d6093bc968a710da, f2c4064ecb76bca69d4ce3fbc52209992f2334dc and 3ad229db0b0ed8538eadcdc2a1c3faf964f239be
